barretenberg/cpp/src/barretenberg/stdlib/recursion/honk/verifier/protogalaxy_recursive_verifier.cpp

```diff
@@ -0,0 +1,305 @@
+    // The degree of K(X) is dk - k - 1 = k(d - 1) - 1. Hence we need  k(d - 1) evaluations to represent it.
+    std::array<FF, VerifierInstances::BATCHED_EXTENDED_LENGTH - VerifierInstances::NUM> combiner_quotient_evals;
+    for (size_t idx = 0; idx < VerifierInstances::BATCHED_EXTENDED_LENGTH - VerifierInstances::NUM; idx++) {
+        combiner_quotient_evals[idx] = transcript->template receive_from_prover<FF>(
+            "combiner_quotient_" + std::to_string(idx + VerifierInstances::NUM));
+    }
+    Univariate<FF, VerifierInstances::BATCHED_EXTENDED_LENGTH, VerifierInstances::NUM> combiner_quotient(
+        combiner_quotient_evals);
+    FF combiner_challenge = transcript->get_challenge("combiner_quotient_challenge");
+    auto combiner_quotient_at_challenge = combiner_quotient.evaluate(combiner_challenge); // fine recursive i think
+
+    auto vanishing_polynomial_at_challenge = combiner_challenge * (combiner_challenge - FF(1));
+    auto lagranges = std::vector<FF>{ FF(1) - combiner_challenge, combiner_challenge };
+
+    // Compute next folding parameters and verify against the ones received from the prover
+    auto expected_next_target_sum =
+        perturbator_at_challenge * lagranges[0] + vanishing_polynomial_at_challenge * combiner_quotient_at_challenge;
+    auto next_target_sum = transcript->template receive_from_prover<FF>("next_target_sum");
+    zero.assert_equal(expected_next_target_sum - next_target_sum, "next target sum mismatch");
+
+    auto expected_betas_star = update_gate_challenges(perturbator_challenge, accumulator->gate_challenges, deltas);
+    for (size_t idx = 0; idx < accumulator->log_instance_size; idx++) {
+        auto beta_star = transcript->template receive_from_prover<FF>("next_gate_challenge_" + std::to_string(idx));
+        zero.assert_equal(beta_star - expected_betas_star[idx],
+                          " next gate challenge mismatch at: " + std::to_string(idx));
+    }
+
+    // Compute ϕ and verify against the data received from the prover
+    WitnessCommitments acc_witness_commitments;
+    auto witness_labels = commitment_labels.get_witness();
+    size_t comm_idx = 0;
+    auto random_generator = Commitment::from_witness(builder, AffineElementNative(ElementNative::random_element()));
+    for (auto& expected_comm : acc_witness_commitments.get_all()) {
+        expected_comm = random_generator;
+        size_t inst = 0;
+        for (auto& instance : instances) {
+            expected_comm = expected_comm + instance->witness_commitments.get_all()[comm_idx] * lagranges[inst];
+            inst++;
+        }
+        auto comm = transcript->template receive_from_prover<Commitment>("next_" + witness_labels[comm_idx]);
+        auto res = expected_comm - comm;
+        random_generator.x.assert_equal(res.x);
+        random_generator.y.assert_equal(res.y);
+        comm_idx++;
+    }
+
+    std::vector<FF> folded_public_inputs(instances[0]->public_inputs.size(), 0);
+    size_t el_idx = 0;
```

Comment:
I'm assuming "el" is some bit of notation from the paper? I found the variable naming a bit confusing here

Reply:
Oh, `el` stands for element in the vector of public inputs, sorry. I'll rename to `public_input` instead